test_conformance/spirv_new/test_decorate.cpp

```diff
@@ -370,60 +446,77 @@ int test_fp_rounding(cl_device_id deviceID,
     return 0;
 }
 
-template<typename Ti, typename To>
-inline To round_to_zero(Ti in)
+template <typename T> static inline double to_double(T in) { return in; }
+
+template <> inline double to_double(cl_half in) { return cl_half_to_float(in); }
+
+template <typename Ti, typename To> static inline To round_to_zero(Ti in)
 {
-    To out = (To)(in);
-    return out;
+    return (To)to_double(in);
 }
 
-template<typename T>
-int sign(T val)
+template <typename T> static inline int sign(T val)
 {
     if (val < 0) return -1;
     if (val > 0) return 1;
     return 0;
 }
 
-template<typename Ti, typename To>
-inline To round_to_even(Ti in)
+template <typename Ti, typename To> static inline To round_to_even(Ti in)
 {
-    // https://en.wikipedia.org/wiki/Rounding#Round_half_to_even
-    return std::floor(in + 0.5) - 1 + std::abs(sign(reference_remainderl((long double)in, 2) - 0.5));
+    double din = to_double(in);
+    return std::floor(din + 0.5) - 1
+        + std::abs(sign(reference_remainderl((long double)din, 2) - 0.5));
 }
 
-template<typename Ti, typename To>
-inline To round_to_posinf(Ti in)
+template <typename Ti, typename To> static inline To round_to_posinf(Ti in)
 {
-    To out = std::ceil(in);
-    return out;
+    return std::ceil(to_double(in));
 }
 
-template<typename Ti, typename To>
-inline To round_to_neginf(Ti in)
+template <typename Ti, typename To> static inline To round_to_neginf(Ti in)
+{
+    return std::floor(to_double(in));
+}
+
+template <typename Ti, typename To>
+static inline Ti generate_fprounding_input(RandomSeed &seed)
 {
-    To out = std::floor(in);
-    return out;
+    constexpr auto minVal = std::numeric_limits<To>::min() / 2;
+    constexpr auto maxVal = std::numeric_limits<To>::max() / 2;
+
+    if (std::is_same<cl_half, Ti>::value)
+    {
+        cl_float f = genrandReal_range<cl_float>(minVal, maxVal, seed);
+        return cl_half_from_float(f, CL_HALF_RTE);
+    }
+
+    return genrandReal_range<Ti>(minVal, maxVal, seed);
```

Comment:
Proposed fix for Clang implicit conversion warning:

```suggestion
    if (std::is_same<cl_half, Ti>::value)
    {
        constexpr auto minVal =
            static_cast<cl_float>(std::numeric_limits<To>::min() / 2);
        constexpr auto maxVal =
            static_cast<cl_float>(std::numeric_limits<To>::max() / 2);
        cl_float f = genrandReal_range<cl_float>(minVal, maxVal, seed);
        return cl_half_from_float(f, CL_HALF_RTE);
    }

    constexpr auto minVal = static_cast<Ti>(std::numeric_limits<To>::min() / 2);
    constexpr auto maxVal = static_cast<Ti>(std::numeric_limits<To>::max() / 2);
    return genrandReal_range<Ti>(minVal, maxVal, seed);
```